src/query_plan/operation.rs

```diff
@@ -2776,6 +2965,49 @@ impl NormalizedFieldSelection {
     pub(crate) fn has_defer(&self) -> bool {
         self.field.has_defer() || self.selection_set.as_ref().is_some_and(|s| s.has_defer())
     }
+
+    fn can_add_to(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        schema: &ValidFederationSchema,
+    ) -> Result<bool, FederationError> {
+        match (parent_type, &self.field.data().field_position) {
+            (
+                CompositeTypeDefinitionPosition::Object(ty_a),
+                FieldDefinitionPosition::Object(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            (
+                CompositeTypeDefinitionPosition::Interface(ty_a),
+                FieldDefinitionPosition::Interface(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            (
+                CompositeTypeDefinitionPosition::Union(ty_a),
+                FieldDefinitionPosition::Union(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            _ => {}
+        }
+
+        let Some(ty) = self.field.type_if_added_to(parent_type, schema)? else {
+            return Ok(false);
+        };
+
+        if let Some(set) = &self.selection_set {
+            if set.type_position != ty {
+                if ty.is_object_type() || ty.is_interface_type() {
```

Comment:
I think this should should also check for union?

Reply:
Hmm.. after looking at this again, I don't think this check is needed. The [original line](https://github.com/apollographql/federation/blob/bbf83946800d0e6f83307e1941cd7052f8a3c38d/internals-js/src/operations.ts#L3103) checks for composite types (which I got my wired crossed with). However, `ty` is an `CompositeTypeDefinitionPosition` and can only be a compose type.